README.md
# Promises Extension

[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![Build Status](https://travis-ci.com/gkampitakis/promises-extension.svg?branch=master)](https://travis-ci.com/gkampitakis/promises-extension)

## Description

A module containing some useful methods for extending native Promises.

## Usage

```javascript
//Using Node.js `require()`
const PromisesExtension = require('promises-extension');

//Using ES6 import
import * as PromisesExtension from 'promises-extension';
```

## Overview

### PromisesExtension.props

Accepts an object containing fields with async code and returns an object with resolved values.

> Supports nested object with promises.

```javascript
const delay = (time, item) =>
    new Promise((resolve: any) => setTimeout(resolve(item), time));

const asyncObject = {
    field: delay(60000, {
        nested: {
            field: 'string',
        },
    }),
    asyncField: fetch(
        'https://jsonplaceholder.typicode.com/todos/1'
    ).then((response) => response.json()),
    normalField: {
        array: ['123', '13'],
    },
};

PromisesExtension.props(asyncObject).then((object) => {
    console.log(object);
});
```

### PromisesExtension.map

Accepts an array and an async callback function and returns a resolved array.

```javascript
const urlArray = [
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
];

PromisesExtension.map(
    urlArray,
    (url) => fetch(url).then((response) => response.json()),
    {
        concurrency: 2,
    }
).then((result) => console.log(result));
```

### PromisesExtension.each

Accepts an array and an async callback function which can be called with 3 parameters `value,index.arrayLength` and returns the array resolved.

```javascript
PromisesExtension.each(
    ['1', Promise.resolve('2'), 3, delay(5000, '4')],
    (value, index, length) => {
        console.log(value, index, length);
    }
).then((res) => console.log(res));
```

### PromisesExtension.delay

Accepts an integer number and an optional callback async function. Returns a promise that will resolve after the a delay in the specified time in ms.

```javascript
PromisesExtension.delay(1000, () => Promise.resolve('1')).then((result) => {
    console.log(result);
});
```

It can also have an optional boolean parameter

```javascript
PromisesExtension.delay(1000, true, () => Promise.resolve('1')).then(
    (result) => {
        console.log(result);
    }
);
```

which determines the order in which the delay is going to take place. If the parameter is `true` first the callback function is executed
and then the delay happens as if the parameter is false first waits for the delay and then executes the function.

> The default behavior is first wait for the delay and then execute the function.

### Promises.allSettled

A polyfill for native Promise.allSettled introduced at 12.10 version.

### Changelog

[CHANGELOG.md](./CHANGELOG.md)

### Example

You can also check an [example](./example) usage.

### Issues

For any [issues](https://github.com/gkampitakis/promises-extension/issues).

## License 

MIT License

CHANGELOG.md
## [2.0.1](https://github.com/gkampitakis/promises-extension/compare/v2.0.0...v2.0.1) (2020-11-14)


### Bug Fixes

* add MIT license ([44dcef8](https://github.com/gkampitakis/promises-extension/commit/44dcef89f6755fb0aec815aefe878b675ff6a669))

# [2.0.0](https://github.com/gkampitakis/promises-extension/compare/v1.0.1...v2.0.0) (2020-10-11)


### Bug Fixes

* change default export ([e2fea2a](https://github.com/gkampitakis/promises-extension/commit/e2fea2aebd2b1c5e24a272e9408f236e4a0387d5))
* performance updates and minor issues ([4ed257e](https://github.com/gkampitakis/promises-extension/commit/4ed257e1ed03e1b477979dc6182751a2e9a91e5d))
* props simple values not resolving ([081e080](https://github.com/gkampitakis/promises-extension/commit/081e0800a1cfc0820d8a84cfe737fd1a40fa84b2))


### BREAKING CHANGES

* update to simple exports

## [1.0.1](https://github.com/gkampitakis/promises-extension/compare/v1.0.0...v1.0.1) (2020-09-16)


### Bug Fixes

* misconfigured export and add support for any value in allSettled ([de9339e](https://github.com/gkampitakis/promises-extension/commit/de9339e02cf16c37a29d88ec78837fddd344744a))

# 1.0.0 (2020-09-16)


### Bug Fixes

* fix error of exposing wrong file ([3a2caeb](https://github.com/gkampitakis/promises-extension/commit/3a2caeb865eefbab91ef7c3314cde1325b0bf2d3))
* update dependencies ([b480492](https://github.com/gkampitakis/promises-extension/commit/b480492f7ebb5c6c289eb6d6dcbbe711a9dbd7e4))
* update map method options to match with bluebird ([b436bdb](https://github.com/gkampitakis/promises-extension/commit/b436bdb9d817a7311685e8a8ebcd3a4299b357fb))


### Features

* **new:** add support on props for nested objects ([eb80547](https://github.com/gkampitakis/promises-extension/commit/eb8054731ff7c5d58f6435678ebb02c12785e2dd))
* **new:** extend delay to support when the delay is taking place ([408af03](https://github.com/gkampitakis/promises-extension/commit/408af03c4103735f53e611ae6f13849d02bf279d))
* **new:** implement delay method ([94a68dc](https://github.com/gkampitakis/promises-extension/commit/94a68dc842f229379ee38c0e17101a7896e7e706))
* **new:** implement each method ([2424212](https://github.com/gkampitakis/promises-extension/commit/24242121613df9e255c7dbf4b1dd99ea551fb77b))
* add new structure and start building util functions ([a2e56c7](https://github.com/gkampitakis/promises-extension/commit/a2e56c759432798066fd72114afaf2897f73f852))
* implement promise.map method ([cd3271f](https://github.com/gkampitakis/promises-extension/commit/cd3271f69f81767b6aeaba8d9ae6e1a82411d8ab))
* migrate from js to ts ([7e30a92](https://github.com/gkampitakis/promises-extension/commit/7e30a92065273eed7fc00f217e52c686edec119c))


example/example.ts
import * as PromisesExtension from '../src/PromisesExtension';
import fetch from 'node-fetch';

const delay = (time, item: any = undefined) =>
  new Promise((resolve: any) => setTimeout(resolve(item), time));

async function Main () {
  const asyncObject = {
    test: delay(60000, {
      hello: {
        test: 'string',
      },
    }),
    test2: fetch(
      'https://jsonplaceholder.typicode.com/todos/1',
      {}
    ).then((response) => response.json()),
    normal: {
      array: ['123', '13'],
    },
  };

  PromisesExtension.props(asyncObject).then((object) => {
    console.log(object);
  });

  const urlArray = [
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
    'https://jsonplaceholder.typicode.com/todos/1',
  ];

  PromisesExtension.map(
    urlArray,
    (url) => fetch(url, {}).then((response) => response.json()),
    {
      concurrency: 2,
    }
  ).then((result) => console.log(result));

  PromisesExtension.each(
    ['1', Promise.resolve('2'), 3, delay(5000, '4')],
    (value, index, length) => {
      console.log(value, index, length);
    }
  ).then((res) => console.log(res));

  PromisesExtension.delay(1000, false, () =>
    fetch('https://jsonplaceholder.typicode.com/todos/1', {}).then((response) =>
      response.json()
    )
  ).then((result) => {
    console.log(result);
  });
}

Main();
